GUIDELINES.md

```diff
@@ -0,0 +1,892 @@
+
+    fun whenLoggingIn(email: String, password: String) {
+        viewModel.loginOrRegister(email, password)
+    }
+
+// Not needed anymore:
+
+//  fun givenExistingUser(email: String, password: String, authToken: AuthToken) {
+//      `when`(backendGateway.checkEmail(email)).thenReturn(true)
+//      `when`(backendGateway.login(email, password)).thenReturn(authToken)
+//  }
+
+
+//  fun thenEmailWasCheckedAtBackend(email: String) {
+//      verify(backendGateway).checkEmail(email)
+//  }
+
+}
+```
+As you can see, all the details regarding how `AuthManager` is being tested can go to the `AuthManagerSteps`, including configuration.
+
+On top of that we add access to `backend` and again offer it as a public member so it can be used in the test. Also we use `BackendFakeUser.USER_A` in order to not only reuse test code and configuration but also test _data_! Here's the result:
+
+```kotlin
+class LoginTest : BaseJUnitTest(appModuleTestingConfiguration) {
+
+    private val sut by steps<LoginSteps>()
+
+    @Test
+    fun `Logging in checks email at backend`() {
+        sut {
+            backend.givenExistingUser(USER_A)
+            whenLoggingIn(USER_A.email, USER_A.password)
+            backend.thenEmailWasChecked(USER_A.email)
+        }
+    }
+
+// Before:
+
+//  @Test
+//  fun `Logging in checks email at backend`() {
+//      sut {
+//          givenExistingUser(email = EXISTING_EMAIL, password = EXISTING_PASSWORD, authToken = EXISTING_AUTH_TOKEN)
+//          whenLoggingIn(email = EXISTING_EMAIL, password = EXISTING_PASSWORD)
+//          thenEmailWasCheckedAtBackend(EXISTING_EMAIL)
+//      }
+//  }
+
+// Not needed anymore:
+
+//  companion object {
+//      const val EXISTING_EMAIL = "existing@test.com"
+//      const val EXISTING_PASSWORD = "supersecure1"
+//      const val EXISTING_AUTH_TOKEN = "auth_token"
+//  }
+    
+}
+```
+
+### Summing up
+
+In this journey it becomes apparent that having integration tests can be simplified a lot when there is already a steps class that tests a smaller unit (`AuthManagerSteps`): it basically requires just using (`... by steps<AuthManagerSteps>()`) or including (`requireSteps<AuthManagerSteps>`) the respective steps class and as a result instead of using a mock there is suddenly a real instance of the class (`AuthManager`) in the test system with just very low effort.
+
+The fake backend example also shows how code reuse can work very easily. In this example we were able to reduce the test code by more or less a half.
+
+All this of course comes with a cost: steps classes have to be set up and very well thought through and preferably be modelled around abstract business models (`BackendFakeSteps`, `LoginSteps`), but that's not always possible or even reasonable (e.g. the `AuthManagerSteps` is a direct reference to a class, which is OK as it serves as an abstraction layer in two test classes).
```

Comment:
This might also be in an extra section. Instead of adding this to summing up, it could be in something like `Things to look out` and laid out in bullet points

Reply:
after implementing your suggestion i think this is somehow obsolete?